sdk/core/core/src/az_http_request_builder.c

```diff
@@ -12,84 +12,231 @@
+az_const_span const AZ_HTTP_METHOD_VERB_HEAD = AZ_CONST_STR("HEAD");
+az_const_span const AZ_HTTP_METHOD_VERB_POST = AZ_CONST_STR("POST");
+az_const_span const AZ_HTTP_METHOD_VERB_PUT = AZ_CONST_STR("PUT");
+az_const_span const AZ_HTTP_METHOD_VERB_DELETE = AZ_CONST_STR("DELETE");
+az_const_span const AZ_HTTP_METHOD_VERB_TRACE = AZ_CONST_STR("TRACE");
+az_const_span const AZ_HTTP_METHOD_VERB_OPTIONS = AZ_CONST_STR("OPTIONS");
+az_const_span const AZ_HTTP_METHOD_VERB_CONNECT = AZ_CONST_STR("CONNECT");
+az_const_span const AZ_HTTP_METHOD_VERB_PATCH = AZ_CONST_STR("PATCH");
+
 az_result az_http_request_builder_init(
     az_http_request_builder * const p_hrb,
     az_span const buffer,
-    int16_t const max_url_size,
     az_const_span const method_verb,
-    az_const_span const initial_url) {
-  (void)p_hrb;
-  (void)buffer;
-  (void)max_url_size;
-  (void)method_verb;
-  (void)initial_url;
+    az_const_span const initial_url,
+    uint16_t const max_url_size,
+    uint16_t const max_headers) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  size_t const method_verb_and_url_size = method_verb.size + max_url_size;
+  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(method_verb)
+      || !az_const_span_is_valid(initial_url) || max_url_size < initial_url.size
+      || method_verb_and_url_size < method_verb.size || method_verb_and_url_size < max_url_size
+      || az_const_span_is_overlap(az_span_to_const_span(buffer), method_verb)
+      || az_const_span_is_overlap(az_span_to_const_span(buffer), initial_url)) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < method_verb.size + max_url_size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0'));
+
+  az_span method_verb_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy(
+      (az_span){ .begin = buffer.begin, .size = method_verb.size }, method_verb, &method_verb_buf));
+
+  az_span uri_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy(
+      (az_span){ .begin = method_verb_buf.begin + method_verb.size, .size = max_url_size },
+      initial_url,
+      &uri_buf));
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = az_span_to_const_span(method_verb_buf),
+                                      .url = uri_buf,
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
-/**
- * @brief
- *
- * @param p_hrb
- * @param name
- * @param value
- * @return az_result
- */
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
```

Comment:
Could we split this code to some smaller functions or use code blocks `{}`. For me, it looks very complicated.

Reply:
I'd love to see comments in here. What is the logic (algorithm) that you're using to accomplish this task. I'd do something like "// First determine is the passed-in query parameter is already in the URL" then have the code for this. 
Then, "// We found the passed-in QP, lets remove it and its value", then have the code this.

Then, "// Append the passed-in QP and its value to the end of the URL", then have the code for this.

Anyway, I write code like this and I always write the comments first. This way I think about the big picture, divide it into smaller tasks/concepts, and then write the code to execute each concept. Then, (me and) other engineers can read/maintain the code to understand the main way of attacking the problem and then see how I went about implementing it. If there is a bug, then another engineer can understand the attempted approach to see why the bug happened and how to fix it.